src/core/processing/queue.ts

```diff
@@ -0,0 +1,265 @@
+ * When a tasks completes with failure (`reject`s or exceeds the timeout), the
+ * respective handler (`catchError` or `catchTimeout`) is called.
+ *
+ * The decaying deque has its name from the observation that new elements are
+ * appended to the tail, and the old elements are removed at arbitrary positions
+ * in the queue whenever a task completes, hence, the queue seems to _decay_.
+ */
+export class DecayingDeque<E> {
+  /**
+   * Number of drifts in the queue. Equivalent to the number of currently
+   * pending tasks.
+   */
+  private length: number = 0
+  /** Head element (oldest), `null` iff the queue is empty */
+  private head: Drift<E> | null = null
+  /** Tail element (newest), `null` iff the queue is empty */
+  private tail: Drift<E> | null = null
+
+  /**
+   * Number of currently pending tasks that we strive for (`add` calls will
+   * resolve only after the number of pending tasks falls below this value.
+   *
+   * In the context of `telegraf`, it is possible to `await` calls to `add` to
+   * determine when to fetch more updates.
+   */
+  private readonly concurrency: number
+  /**
+   * Timer that waits for the head element to time out, will be rescheduled
+   * whenever the head element changes. It is `undefined` iff queue is empty.
+   */
+  private timer: NodeJS.Timeout | undefined
+  /**
+   * List of subscribers that wait for the queue to be empty. All functions in
+   * this array will be called as soon as new capacity is available, i.e. the
+   * number of pending tasks falls below `concurrency`.
+   */
+  private subscribers: Array<(capacity: number) => void> = []
+
+  /**
+   * Creates a new decaying queue with the given parameters.
+   *
+   * @param handlerTimeout max period of time for a task
+   * @param worker task generator
+   * @param concurrency `add` will return only after the number of pending tasks fell below `concurrency`. `false` means `1`, `true` means `Infinity`, numbers below `1` mean `1`
+   * @param catchError error handler, receives the error and the source element
+   * @param catchTimeout timeout handler, receives the source element and the promise of the task
+   */
+  constructor(
+    private readonly handlerTimeout: number,
+    private readonly worker: (t: E) => Promise<void>,
+    concurrency: boolean | number,
+    private readonly catchError: (err: unknown, elem: E) => Promise<void>,
+    private readonly catchTimeout: (t: E, task: Promise<void>) => void
+  ) {
+    if (concurrency === false) this.concurrency = 1
+    else if (concurrency === true) this.concurrency = Infinity
+    else this.concurrency = concurrency < 1 ? 1 : concurrency
+  }
+
+  /**
+   * Adds the provided elements to the queue and starts tasks for all of them
+   * immediately. Returns a `Promise` that resolves with `concurrency - length`
+   * once this value becomes positive.
+   * @param elems elements to be added
+   * @returns `capacity - length` inside a `Promise`
+   */
+  add(elems: E[]): Promise<number> {
```

Comment:
1. Wouldn't this be simpler if it accepted one element at a time?
2. It would be way simpler if you dropped all the list stuff, Node seems to use a list internally: https://stackoverflow.com/a/13616530

Reply:
On 2nd thought, you cannot drop list stuff, it's needed for the custom backpressure.

Am I correct in saying that backpressure is the crucial difference from #1262? The face that `add` settles when bot is "bored", rather than when when middleware settles.

Reply:
Yes, that and the fact that the number of polled updates dynamically adjusts to the load. The higher the load, the updates are requested.

Also, errors are handled after timeout instead of being swallowed.

I do have more improvements in mind, e.g. providing a way to give updates more time upon certain kinds or errors. Might help with one bot I have that has comparably long-running middleware due to expensive operations, and would have been trivial to implement with this PR

Reply:
Also, we don't use timers, but `unref`ed timers. I know that timers are fast, but the docs explicitly state that one should avoid `unref`fing too many. We'd `unref` up to 1000/s or so at peak loads.

I tried to figure out in the Node.js source why this is supposed to be slow and if we can ignore this advice, but I could not find out about this. (It is not a coincidence that I used a comparable data structure than the internal one in Node, updates are handled similar to timeouts.) It may be possible to rewrite the rest without using an explicit queue (effectively leveraging the event loop as queue in a way), but as I figured that we need it anyway, I didn't investigate it at the time.

Reply:
On 3rd thought, all we need for "improved" backpressure is number of updates currently being processed, no need for list.

> providing a way to give updates more time upon certain kinds or errors.

Interesting, but again detrimental for polling correctness.

Reply:
> no need for list

That's what I meant with my comment above and with https://github.com/telegraf/telegraf/pull/1262#issuecomment-752932119, I'll investigate this approach in the coming days.

> detrimental for polling correctness

This may be a nitpick, but there's a difference between correctness and reliability. If unreliability is desired for other benefits, it may be the correct thing to do (example: UDP). (We agree that this should be opt-in and not the default.)